### PR TITLE
Fix: Type hinting on Application.add_handlers

### LINKED ISSUE #4530: [Type Hinting] Application.add_handlers is not properly annotated.
### Steps to Reproduce

1. Create a main.py file like:
```python
from telegram import Update
from telegram.ext import (
    ApplicationBuilder,
    CommandHandler,
    ContextTypes,
    MessageHandler,
    filters,
)

# pyright: strict


async def hello(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    if update.message and update.effective_user:
        await update.message.reply_text(f"Hello {update.effective_user.first_name}")


async def echo(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    if update.message and update.message.text:
        await update.message.reply_text(update.message.text)


async def hi(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    if update.message:
        await update.message.reply_text("Hi")


async def welcome(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    if update.message:
        await update.message.reply_text("Welcome")


mixed_handlers = [
    CommandHandler("hello", hello),
    MessageHandler(filters.TEXT & ~filters.COMMAND, echo),
]


command_handlers = [
    CommandHandler("hi", hi),
    CommandHandler("welcome", welcome),
]


app = ApplicationBuilder().token("YOUR TOKEN HERE").build()


app.add_handlers(mixed_handlers)

app.add_handlers(command_handlers)


app.run_polling()
```
2. Run pyright and mypy on the file:
```bash
 pyright main.py && mypy main.py
```

### Expected behaviour

Code is properly annotated so pyright and mypy doesn't throw any error.

### Actual behaviour

Both mypy and pyright in strict mode will complain about invariance of https://github.com/python-telegram-bot/python-telegram-bot/blob/06854633ab00d81fa5592adc17d92896fbda662c/telegram/ext/_application.py#L1423

Errors:

```
1. Argument 1 to "add_handlers" of "Application" has incompatible type "list[CommandHandler[CallbackContext[ExtBot[None], dict[Any, Any], dict[Any, Any], dict[Any, Any]], None]]"; expected "list[BaseHandler[Any, CallbackContext[ExtBot[None], dict[Any, Any], dict[Any, Any], dict[Any, Any]], Any]] | tuple[BaseHandler[Any, CallbackContext[ExtBot[None], dict[Any, Any], dict[Any, Any], dict[Any, Any]], Any]] | dict[int, list[BaseHandler[Any, CallbackContext[ExtBot[None], dict[Any, Any], dict[Any, Any], dict[Any, Any]], Any]] | tuple[BaseHandler[Any, CallbackContext[ExtBot[None], dict[Any, Any], dict[Any, Any], dict[Any, Any]], Any]]]"  [arg-type]

2. "List" is invariant -- see https://mypy.readthedocs.io/en/stable/common_issues.html#variance

3. Consider using "Sequence" instead, which is covariant

4. Argument 1 to "add_handlers" of "Application" has incompatible type "list[CommandHandler[CallbackContext[ExtBot[None], dict[Any, Any], dict[Any, Any], dict[Any, Any]], None]]"; expected "list[BaseHandler[Any, CallbackContext[ExtBot[None], dict[Any, Any], dict[Any, Any], dict[Any, Any]], Any]] | tuple[BaseHandler[Any, CallbackContext[ExtBot[None], dict[Any, Any], dict[Any, Any], dict[Any, Any]], Any]] | dict[int, list[BaseHandler[Any, CallbackContext[ExtBot[None], dict[Any, Any], dict[Any, Any], dict[Any, Any]], Any]] | tuple[BaseHandler[Any, CallbackContext[ExtBot[None], dict[Any, Any], dict[Any, Any], dict[Any, Any]], Any]]]"  [arg-type]

5. Argument of type "list[CommandHandler[DEFAULT_TYPE, None]]" cannot be assigned to parameter "handlers" of type "List[BaseHandler[Any, DEFAULT_TYPE, Any]] | Tuple[BaseHandler[Any, DEFAULT_TYPE, Any]] | Dict[int, List[BaseHandler[Any, DEFAULT_TYPE, Any]] | Tuple[BaseHandler[Any, DEFAULT_TYPE, Any]]]" in function "add_handlers"
Type "list[CommandHandler[DEFAULT_TYPE, None]]" is not assignable to type "List[BaseHandler[Any, DEFAULT_TYPE, Any]] | Tuple[BaseHandler[Any, DEFAULT_TYPE, Any]] | Dict[int, List[BaseHandler[Any, DEFAULT_TYPE, Any]] | Tuple[BaseHandler[Any, DEFAULT_TYPE, Any]]]"
	"list[CommandHandler[DEFAULT_TYPE, None]]" is not assignable to "List[BaseHandler[Any, DEFAULT_TYPE, Any]]"
Type parameter "_T@list" is invariant, but "CommandHandler[DEFAULT_TYPE, None]" is not the same as "BaseHandler[Any, DEFAULT_TYPE, Any]"
Consider switching from "list" to "Sequence" which is covariant
"list[CommandHandler[DEFAULT_TYPE, None]]" is not assignable to "Tuple[BaseHandler[Any, DEFAULT_TYPE, Any]]"
"list[CommandHandler[DEFAULT_TYPE, None]]" is not assignable to "Dict[int, List[BaseHandler[Any, DEFAULT_TYPE, Any]] | Tuple[BaseHandler[Any, DEFAULT_TYPE, Any]]]" [reportArgumentType]
```

### Operating System

Arch Linux

### Version of Python, python-telegram-bot & dependencies

python 3.12.7

python-telegram-bot 21.6

mypy 1.12.1

pyright 1.1.385

### Relevant log output

_No response_

### Additional Context

Changing List[BaseHandler[Any, CCT, Any] to Sequence[BaseHandler[Any, CCT, Any] seems to fix the issue.

```diff
     def add_handlers(
        self,
        handlers: Union[
-          Union[List[BaseHandler[Any, CCT, Any]], Tuple[BaseHandler[Any, CCT, Any]]],
+          Sequence[BaseHandler[Any, CCT, Any]],
            Dict[int, Union[List[BaseHandler[Any, CCT, Any]], Tuple[BaseHandler[Any, CCT, Any]]]],
        ],
        group: Union[int, DefaultValue[int]] = _DEFAULT_0,
    ) -> None:

```

### PR DESCRIPTION
**Pull request by roast-lord**
Possible fix #4530

### CODE REVIEW

**Review by Bibo-Joshi — CHANGES_REQUESTED** [review]
Hi. Thanks for the catch and also for already providing a PR!
The annotation change itself looks ok for me. However, both the docstring and method implementation will need adaption as well. See lines 1435, 1463 and 1467 (1467 doesn't need adaption, but explains why adaption of 1463 is needed). Please be sure to adapt/extend unit tests if required. I'd say this change also warrants a "versionchanged" directive. Please have a look at the PR-checklist that we link in our PR-Template :)

**Comment by roast-lord** [comment]
Hi! Thanks for the feedback!

I've made the changes you asked, let me know if there's something wrong!

File: telegram/ext/_application.py [hunk 1]
```diff
@@ -1431,12 +1432,15 @@ def add_handlers(
         .. versionadded:: 20.0
 
         Args:
-            handlers (List[:class:`telegram.ext.BaseHandler`] | \
-                Dict[int, List[:class:`telegram.ext.BaseHandler`]]): \
+            handlers (Sequence[:class:`telegram.ext.BaseHandler`] | \
+                Dict[int, List[:class:`telegram.ext.BaseHandler`] | \
+                Tuple[:class:`telegram.ext.BaseHandler`]]):
```
Comment (Bibo-Joshi) [inline telegram/ext/_application.py:1437]:
```suggestion
                Dict[int, Sequence[:class:`telegram.ext.BaseHandler`]]):
```
should this be updated as well?
Do you get any type checker problems when passing a dictionary?

File: telegram/ext/_application.py [hunk 2]
```diff
@@ -1431,12 +1432,15 @@ def add_handlers(
...
                 Specify a sequence of handlers *or* a dictionary where the keys are groups and
                 values are handlers.
-            group (:obj:`int`, optional): Specify which group the sequence of :paramref:`handlers`
-                should be added to. Defaults to ``0``.
```
Comment (Bibo-Joshi) [inline telegram/ext/_application.py:1439]: This should not be removed :)

**Review by Bibo-Joshi — CHANGES_REQUESTED** [review]

File: telegram/ext/_application.py (hunk 1 above)
Comment (roast-lord) [reply telegram/ext/_application.py:1437]: No I did not get any type checker problems.

I just updated this documentation part to match the actual function type hint.

File: telegram/ext/_application.py (hunk 2 above)
Comment (roast-lord) [reply telegram/ext/_application.py:1439]: Oops, this was a mistake, not intentional 😅.

File: telegram/ext/_application.py (hunk 1 above)
Comment (Bibo-Joshi) [reply telegram/ext/_application.py:1437]: mh, using 

```python
app.add_handlers({1: mixed_handlers})

app.add_handlers({1: command_handlers})
```

in your MWE in #4530, I get 

```shell
mypy foo.py
foo.py:50: error: Dict entry 0 has incompatible type "int": "List[CommandHandler[CallbackContext[ExtBot[None], Dict[Any, Any], Dict[Any, Any], Dict[Any, Any]], None]]"; expected "int": "Union[List[BaseHandler[Any, CallbackContext[ExtBot[None], Dict[Any, Any], Dict[Any, Any], Dict[Any, Any]], Any]], Tuple[BaseHandler[Any, CallbackContext[ExtBot[None], Dict[Any, Any], Dict[Any, Any], Dict[Any, Any]], Any]]]"  [dict-item]
Found 1 error in 1 file (checked 1 source file)
```

changing the type hint for the dict values to sequence fixes the problem. So I'd suggest to update that as well

Comment (roast-lord) [reply telegram/ext/_application.py:1437]: Oh...

When I tested I added to the MWE this:

```python
app.add_handlers(
    {
        1: [
            CommandHandler("hello", hello),
            MessageHandler(filters.TEXT & ~filters.COMMAND, echo),
        ]
    }
)

app.add_handlers(
    {
        1: [
            CommandHandler("hi", hi),
            CommandHandler("welcome", welcome),
        ]
    }
)

```
This, for some reason, does not trigger any mypy errors.

Comment (roast-lord) [reply telegram/ext/_application.py:1437]: But thanks for the catch too!

Will send a update right away!

Comment (roast-lord) [reply telegram/ext/_application.py]: Done!

File: telegram/ext/_application.py [hunk 3]
```diff
@@ -1431,10 +1431,15 @@ def add_handlers(
...
 
         Args:
-            handlers (List[:class:`telegram.ext.BaseHandler`] | \
-                Dict[int, List[:class:`telegram.ext.BaseHandler`]]): \
+            handlers (Sequence[:class:`telegram.ext.BaseHandler`] | \
+                Dict[int, Sequence[:class:`telegram.ext.BaseHandler`]]):
                 Specify a sequence of handlers *or* a dictionary where the keys are groups and
                 values are handlers.
+
+                .. versionchanged:: 21.7
```
Comment (Bibo-Joshi) [inline telegram/ext/_application.py:1439]:
```suggestion
                .. versionchanged:: NEXT.VERSION
```

**Review by Bibo-Joshi — APPROVED** [review]
LGTM, then. thanks very much for the contribution and the fast responses!

**Pull request merged**